### PR TITLE
Enhance vectorization template and implement generic `Concat` op

### PR DESCRIPTION
**Pull request by Somoku**
## Motivation

This PR introduces the following improvements to the framework:

1. **Vectorization template enhancement**: Existing vectorization templates lack support for variable input/output counts with indexing (required by Dropout/Dropout2d kernels). This PR generalizes template support for arbitrary input/output configurations with index.

2. **Performance optimization of generic `Concat` kernel**: Existing `Concat` kernel exhibits suboptimal performance in multi-input concatenation which is naively composed of sequential `copy` operations. This PR adds optimized `Concat` kernel implementation using batched memory operations, achieving similar performance to PyTorch. The performance results are listed below:

```
Torch Concat with shape  (8192, 1024) , dim  1 : 0.0930908203125
Hetu Concat with shape  (8192, 1024) , dim  1 : 0.09376960000000001
Torch ConcatGradient with shape  (8192, 1024) , dim  1 : 0.1780517578125
Hetu ConcatGradient with shape  (8192, 1024) , dim  1 : 0.08901440000000001
```

3. **Memory optimization in backward ops**: Unnecessary retention of full NDArray storage data (activation) during backward process leads to increased memory consumption. This PR decouples metadata transmission from storage data by leveraging RuntimeContext/InstantiationContext for lightweight parameter passing, enabling early activation release while preserving operational semantics.

## Modifications

- Refactor vectorization templates to support multi-outputs(inputs) with index and optimize Dropout kernel
- Optimize `Concat` kernel and op to support multi-inputs efficiently
- Revamp context store to avoid unnecessary activation dependency
- Support optimizers w/o scheduler to run unit tests

## Known Issues

- When the metadata is migrated from fw op to bw op, the metadata will not be popped from fw op context. This design is due to the requirement of unit tests, which will create backward ops multiple times during profiling.
- Currently we don't support distributed training in eager graph, thus this PR adds some workarounds to ensure the unit-tests will run successfully
- During the optimization of `Concat` kernel, I found that our `OffsetCalculator` implementation may be inefficient for kernels due to the occupancy of more registers. Passing shape and stride arrays into constant memory might be more efficient for kernels leveraging more registers. However, due to my limited knowledge about CUDA kernel optimization, it needs more future exploration and discussion.

### CODE REVIEW

**Comment by ccchengff** [comment]
+2
